feat(theme) add dark mode

edlerd opened the pull request:
## Done

feat(theme) add dark mode

Fixes #1341

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @Kxiru or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](https://github.com/canonical/lxd-ui/blob/main/CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - go to settings and switch the theme to dark mode
    - check if pages and interactions work just as before in dark and light mode.

## Screenshots

![image](https://github.com/user-attachments/assets/0988e7bf-f6a4-4f10-8945-5079796db045)

![image](https://github.com/user-attachments/assets/4907eeda-e1b3-4996-9001-711fa7302501)

![image](https://github.com/user-attachments/assets/e47701c4-7292-479e-b9bd-6d0dd69a9062)

piperdeck commented:
Is this what bar graphs are supposed to look like? Looks a little weird to me, but if this is what the upstream darkmode calls for then I can live with it

![CleanShot 2025-06-30 at 15 28 16@2x](https://github.com/user-attachments/assets/58a47c88-206d-4968-839d-8978eabbd04d)

---

In light mode, we use dark mode to indicate the current navigation of a network diagram. So conversely in dark mode we should use light mode to indicate the navigation (current node in the diagram)

![CleanShot 2025-06-30 at 15 29 21@2x](https://github.com/user-attachments/assets/78853f26-ace1-473e-8fdd-dd7b685a9868)

---

Looks like code snippets are getting a bit messed up

|  Light    |   Dark  |
| --- | --- |
| ![CleanShot 2025-06-30 at 15 37 00@2x](https://github.com/user-attachments/assets/41b67326-025e-40c9-89c9-0175ea764115) | ![CleanShot 2025-06-30 at 15 36 55@2x](https://github.com/user-attachments/assets/6ddc0cba-f408-4fb0-8ba5-c74b7f850fa0) |
| ![CleanShot 2025-06-30 at 15 37 40@2x](https://github.com/user-attachments/assets/b4907ff0-bb07-4cdf-999f-b5fd3ac2417f) | ![CleanShot 2025-06-30 at 15 37 44@2x](https://github.com/user-attachments/assets/ffa9111a-7abb-4069-9eea-ea9957275386) |

piperdeck commented:
Maybe the switch should go over to the left, next to the version info?

edlerd commented:
> Is this what bar graphs are supposed to look like? Looks a little weird to me, but if this is what the upstream darkmode calls for then I can live with it

updated the bar graphs in light mode and also in dark mode, so they rely on themed colors.

> In light mode, we use dark mode to indicate the current navigation of a network diagram. So conversely in dark mode we should use light mode to indicate the navigation (current node in the diagram)

good catch, updated the network topology styles to be theme based.

> Looks like code snippets are getting a bit messed up

this was another override on our side, adjusted it now.

> Maybe the switch should go over to the left, next to the version info?

I experimented with it, using a smaller switch, left aligning it or using a segmented control with "light, dark, system" options. All looked a bit too intrusive in the footer. So I moved the switch to the settings page. I don't think it should be that prominent in our footer.

edlerd commented:
> So if I was a user who had heard that LXD-UI had dark mode now, the first thing I would do is search for "dark mode" in the settings. But "dark" doesn't return any matches, because the name of the setting doesn't contain that string. Is there some way for these configurations to have a hidden field that we can search for alternate terms like this? Or alternatively, search over values somehow?

Good point, I gave the setting a description and made the search also match on description text. That way the query "dark" matches the theme setting.

piperdeck commented:
Can we use this copy instead:

> Set UI to dark theme, light theme, or match the user’s system theme

Other than that, this is good to go

Kxiru commented on src/pages/settings/SettingForm.tsx:
```diff
@@ -52,6 +53,7 @@ const SettingForm: FC<Props> = ({
   const isLokiAuthPassword = configField.key === "loki.auth.password";
   const isSecret = isTrustPassword || isLokiAuthPassword;
   const isClusteredInput = isClustered && configField.scope === "local";
+  const isTheme = configField.key === "user.ui_theme";
```
More descriptive naming? Is this to say whether the current version HAS a theme? Perhaps hasTheme?

Kxiru commented on src/pages/settings/SettingThemeSwitcher.tsx:
```diff
@@ -0,0 +1,67 @@
+  localStorage.setItem(LOCAL_STORAGE_KEY, theme);
+};
+
+export const isDarkTheme = (theme: string) => {
```
Nice function logic. I had to really unpack this one line by line.

Kxiru commented on src/components/ResourceLink.tsx:
```diff
@@ -20,10 +20,8 @@ const ResourceLink: FC<Props> = ({ type, value, to, disabled }) => {
-        <ResourceIcon type={type} />
-        {value}
-      </span>
+      <ResourceIcon type={type} />
```
Thanks for this fix!

Kxiru reviewed (commented):
I made a mock of what a nav-bar toggle could look like as I thought it would be cute and quaint. What do you think? We could have this be a more accessible toggle.

![image](https://github.com/user-attachments/assets/2ce1999f-936b-4ec3-9af5-959fe4b7b98c)
![image](https://github.com/user-attachments/assets/ed4b33d0-e54f-4875-be50-f4d968b2cad0)
![image](https://github.com/user-attachments/assets/96df5f5a-5d42-413c-a380-a87b9cab8370)

Just using the [Switch Component](https://canonical.github.io/react-components/?path=/docs/components-switch--docs). It's definitely quite out of the way to keep going to Settings to change this.


Tested edges of the UI such as:
- Panels
- Confirmation/upload modals
- Inline actions
- Search and filter

1. I think the sub-menus should be in white for higher contrast, accessibility and visibility:
![image](https://github.com/user-attachments/assets/7468d0c9-2b35-412d-8966-cf2bc7e8ac92)

edlerd commented:
> I made a mock of what a nav-bar toggle could look like as I thought it would be cute and quaint. What do you think? We could have this be a more accessible toggle.
> Just using the [Switch Component](https://canonical.github.io/react-components/?path=/docs/components-switch--docs). It's definitely quite out of the way to keep going to Settings to change this.

I tested the switch in an earlier iteration in the status bar, next to the version number. Concerns were raised on supporting the "system" default among the dark/light choices. So the binary switch input won't work for the theme. Hence, the move of the theme picker into the settings. I think that is ok, this is not the most important setting that needs to be accessed frequently. I agree, putting it more prominently would advertise the feature nicely -- though I am not sure if we should. I think putting it into settings is an okay tradeoff.

> 1. I think the sub-menus should be in white for higher contrast, accessibility and visibility:

Good point, though this is an edge case of instance creation, where the navigation is disabled. After picking a base image, the text in the nav is not muted anymore and gets a higher contrast. So I think this works as intended.

Kxiru reviewed (approved):
A nice new feature to the UI. Thank you for this addition.
